### ISSUE TITLE
Design Discussion • Resonate Timer

### ISSUE THREAD

**Issue by dtornow** [resonatehq/resonate#139]
### Feature

**tl;dr**

A Durable Promise should be able to act as a **timer**/**timeout**

**Description**

Many applications require the ability to wait on a **timer**. In the context of the Resonate Server, a timer is represented by a **Durable Promise** that, on timeout, **resolves** itself

**See Also**

See #132 for Resonate Schedules

**Comment by dtornow** [comment]
### Implementation Considerations

Durable Promises have a timeout feature. When a Durable Promise times out, the Durable Promise gets rejected. We need to be able to distinguish a Durable Promise from a Durable Promise representing a timer

**Solution**

Introduce a `kind` tag. Introduce a `Timer` kind.

**Comment by guergabo** [comment]
## Design Proposal 

### System Design 

**Design Considerations:** 
- _accuracy vs performance_:  The guarantees that any timer system can offer are low. They can guarantee that a timer scheduled to fire after t ticks will have its action executed some time after t ticks. So, if time systems will always be "inaccurate", we can and should optimize for performance. 
- _throughput vs latency_: By processing more operations at once in a batch, overall throughput increases compared to handling operations individually. The larger the batch size the more per-operation latency is hurt due to queueing delays.

**Coroutines**
- Scheduler. TBD.
- Executor. TBD. (mark for deletion?)
- Reaper. TBD. 

**Data Structures**
In terms of data structures to queue up scheduled timers in-memory, [this research paper](https://paulcavallaro.com/blog/hashed-and-hierarchical-timing-wheels/) is very popular in the space of timer systems and looks at all our options: 

1. Unordered List 
2. Sorted List 
3. Priority Queues 
4. Timing Wheels*

**Configurations** 
- Wheel sizes 
- Batch sizes
- Workers. TBD. 

**Caching Strategies**
Given that we have a system that is read-heavy, we can optimize for reliability and implement a _write-through caching strategy_ to make sure that the timer persists in the event of failure. The database will be our source of truth, not the cache. 

Depending on the data structure and how timers are scheduled, we can even postpone storing the timer in-memory if it is not within the time bounds of the batch. TBD.

### Database Design 
**Design Considerations:** 
- The system is read-heavy, as a timer is created once by the client and read many times by the scheduler. 

**Read Operations**
- (internally) find all the timers that are scheduled to run for that batch. consider sharding to distribute load across workers. TBD. 

**Write Operations**
- (externally) create a new timer. Stored with columns like ID, firing time, etc. 
- (internally) delete timers after promoting them to promises

**Comment by guergabo** [comment]
Already completed.
